server/src/main/java/com/epam/aidial/core/server/resource/ResourceDescriptor.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    public String getDecodedUrl() {
+        StringBuilder builder = new StringBuilder();
+        builder.append(type.getGroup())
+                .append(PATH_SEPARATOR)
+                .append(bucketName)
+                .append(PATH_SEPARATOR);
+
+        if (!parentFolders.isEmpty()) {
+            String parentPath = String.join(PATH_SEPARATOR, parentFolders);
+            builder.append(parentPath)
+                    .append(PATH_SEPARATOR);
+        }
+
+        if (name != null) {
+            builder.append(name);
+
+            if (isFolder) {
+                builder.append(PATH_SEPARATOR);
+            }
+        }
+
+        return builder.toString();
+    }
+
+    public String getAbsoluteFilePath() {
+        StringBuilder builder = new StringBuilder();
+        builder.append(bucketLocation)
+                .append(type.getGroup())
+                .append(PATH_SEPARATOR);
+
+        if (!parentFolders.isEmpty()) {
+            builder.append(getParentPath())
+                    .append(PATH_SEPARATOR);
+        }
+
+        if (name != null) {
+            builder.append(name);
+
+            if (isFolder) {
+                builder.append(PATH_SEPARATOR);
+            }
+        }
+
+        return builder.toString();
+    }
+
+    @Nullable
+    public ResourceDescriptor getParent() {
+        if (parentFolders.isEmpty()) {
+            return null;
+        }
+
+        String parentFolderName = parentFolders.get(parentFolders.size() - 1);
+        return new ResourceDescriptor(type, parentFolderName,
+                parentFolders.subList(0, parentFolders.size() - 1), originalPath, bucketName, bucketLocation, true);
+    }
+
+    public boolean isRootFolder() {
+        return isFolder && name == null;
+    }
+
+    public boolean isPublic() {
+        return bucketLocation.equals(BlobStorageUtil.PUBLIC_LOCATION);
+    }
+
+    public boolean isPrivate() {
+        return !isPublic();
+    }
+
+    public String getParentPath() {
+        return parentFolders.isEmpty() ? null : String.join(PATH_SEPARATOR, parentFolders);
+    }
+
+    public ResourceDescriptor resolveByUrl(String url) {
+        String prefix = type.getGroup() + PATH_SEPARATOR + bucketName  + PATH_SEPARATOR;
+        return resolve(prefix, url);
+    }
+
+    public ResourceDescriptor resolveByPath(String url) {
+        String prefix = bucketLocation + type.getGroup() + PATH_SEPARATOR;
+        return resolve(prefix, url);
+    }
+
+    private ResourceDescriptor resolve(String prefix, String url) {
+        if (!isFolder) {
+            throw new IllegalStateException("Resource must be folder");
+        }
+        if (!url.startsWith(prefix)) {
+            throw new IllegalArgumentException("Incompatible description and absolute path");
+        }
+
+        String relativePath = url.substring(prefix.length());
+
+        List<String> paths = Stream.of(relativePath.split(PATH_SEPARATOR)).map(UrlUtil::decodePath).toList();
```

Comment:
Why do you decode each segment when called for a path?

Reply:
We want to put a decoded path to blob store